README.md
# clojureselect

FIXME: description

## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar clojureselect-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2023 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2023-12-15
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2023-12-15
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/clojureselect/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/clojureselect/compare/0.1.0...0.1.1


test/clojureselect/core_test.clj
(ns clojureselect.core-test
  (:require [clojure.test :refer :all]
            [clojureselect.business-logic :refer :all]))


;***********************************************************
;                     OSNOVNE FUNKCIJE 
;***********************************************************

(deftest rating-test
  (testing "Testing get-ratings function"
  (is (= (get-ratings 1) [{:id 1, :candidate-id 1, :job-id 1, :qualification-id 1, :value 10.0}
                          {:id 2, :candidate-id 1, :job-id 1, :qualification-id 2, :value 10.0}
                          {:id 3, :candidate-id 1, :job-id 1, :qualification-id 3, :value 10.0}]))))

(deftest rating-test2
  (testing "Testing get-ratings function"
    (is (= (get-ratings 2) [{:id 4, :candidate-id 2, :job-id 1, :qualification-id 1, :value 8.0}
                            {:id 5, :candidate-id 2, :job-id 1, :qualification-id 2, :value 9.1}
                            {:id 6, :candidate-id 2, :job-id 1, :qualification-id 3, :value 7.0}]))))

(deftest rating-test3
  (testing "Testing get-ratings function"
    (is (= (get-ratings 180) []))))

(deftest rating-test4
  (testing "Testing get-ratings function"
    (is (= (get-ratings -7) []))))


(deftest candidates-test
  (testing "Testing get-candidates function"
    (is (= (get-candidates 1) [{:id 1, :firstname "Dusan", :lastname "Tavic", :active true, :email "dusantavic1@gmail.com", :status 0, :job-id 1}
                               {:id 2, :firstname "Nenad", :lastname "Panovic", :active true, :email "nenadpann@gmail.com", :status 0, :job-id 1}
                               {:id 4,
                                :firstname "Arsenije",
                                :lastname "Pavlovic",
                                :active true,
                                :email "arsenijee00@yahoo.com",
                                :status 0,
                                :job-id 1}]))))

(deftest candidates-test2
  (testing "Testing get-candidates function"
    (is (= (get-candidates 0) []))))

(deftest candidates-test3
  (testing "Testing get-candidates function"
    (is (= (get-candidates "dusan") []))))


(deftest jobs-criteria-test
  (testing "Testing get-jobs-criteria function"
    (is (= (get-jobs-criteria 1) [{:job-id 1, :qualification-id 1, :ponder 0.5}
                                  {:job-id 1, :qualification-id 2, :ponder 0.3}
                                  {:job-id 1, :qualification-id 3, :ponder 0.2}]))))

(deftest jobs-criteria-test2
  (testing "Testing get-jobs-criteria function"
    (is (= (get-jobs-criteria -10) []))))


(deftest job-name-test
  (testing "Testing get-jobs-name function"
    (is (= (get-jobs-name -40) nil))))

(deftest job-name-test2
  (testing "Testing get-jobs-name function"
    (is (= (get-jobs-name 1) "C# Junior Developer"))))

(deftest ponder-test
  (testing "Testing get-ponder function"
    (is (= (get-ponder 1 1) 0.5))))

(deftest ponder-test2
  (testing "Testing get-ponder function"
    (is (> (get-ponder 1 1) 0))))

(deftest ponder-test3
  (testing "Testing get-ponder function"
    (is (= (get-ponder -1 -1) "error. please insert valid id values"))))

(deftest ponder-test4
  (testing "Testing get-ponder function"
    (is (= (get-ponder -1 1) "error. please insert valid id values"))))


(deftest ponder-test5
  (testing "Testing get-ponder function"
    (is (= (get-ponder 1 -1) "error. please insert valid id values"))))

(deftest ratings-of-criteria-test
  (testing "Testing get-ratings-of-criteria function"
    (is (= (get-ratings-of-criteria 1 1) [{:id 1, :candidate-id 1, :job-id 1, :qualification-id 1, :value 10.0}
                                          {:id 4, :candidate-id 2, :job-id 1, :qualification-id 1, :value 8.0}]))))

(deftest ratings-of-criteria-test2
  (testing "Testing get-ratings-of-criteria function"
    (is (= (get-ratings-of-criteria -1 -1) []))))

(deftest ratings-of-criteria-test3
  (testing "Testing get-ratings-of-criteria function"
    (is (= (get-ratings-of-criteria -1 1) []))))

(deftest ratings-of-criteria-test4
  (testing "Testing get-ratings-of-criteria function"
    (is (= (get-ratings-of-criteria 1 -1) []))))


(deftest ratings-of-job-test
  (testing "Testing get-ratings-of-job function"
    (is (= (get-ratings-of-job -1) []))))


(deftest ratings-of-job-test2
  (testing "Testing get-ratings-of-job function"
    (is (= (get-ratings-of-job 1) [{:id 1, :candidate-id 1, :job-id 1, :qualification-id 1, :value 10.0}
                                   {:id 2, :candidate-id 1, :job-id 1, :qualification-id 2, :value 10.0}
                                   {:id 3, :candidate-id 1, :job-id 1, :qualification-id 3, :value 10.0}
                                   {:id 4, :candidate-id 2, :job-id 1, :qualification-id 1, :value 8.0}
                                   {:id 5, :candidate-id 2, :job-id 1, :qualification-id 2, :value 9.1}
                                   {:id 6, :candidate-id 2, :job-id 1, :qualification-id 3, :value 7.0}]))))


test/clojureselect/business_logic_test.clj
;; (ns clojureselect.business-logic-test
;;   (:require [midje.sweet :refer [fact facts]]
;;             [clojureselect.business-logic :refer :all]))


;; (fact "Test for get-rating function"
;;       (get-ratings 1) => [{:id 1, :candidate-id 1, :job-id 1, :qualification-id 1, :value 10.0}
;;                           {:id 2, :candidate-id 1, :job-id 1, :qualification-id 2, :value 10.0}
;;                           {:id 3, :candidate-id 1, :job-id 1, :qualification-id 3, :value 10.0}])

src/clojureselect/data.clj
(ns clojureselect.data
  (:require [clojureselect.repository :as repository])
  (:gen-class))

;***********************************************************
;            ALATI I METODE VESTACKE INTELIGENCIJE         *
;                I SOFTVERSKOG INZENJERSTVA                *
;                          2023.                           *
;             --------------------------------             *
;                      Clojure Select                      *
;              ~Decision Support Clojure App~              *
;             --------------------------------             *
;                        Dusan Tavic                       *
;                         2023/3801                        *
;***********************************************************


;***********************************************************
;                 SIMULACIJA PODATAKA U BAZI 
;***********************************************************

;; (def candidates [{:id 1
;;                   :firstname "dusan"
;;                   :lastname "tavic"
;;                   :active true
;;                   :email "dusantavic1@gmail.com"
;;                   :status "rated"
;;                   :job-id 1},
;;                  {:id 2
;;                   :firstname "nenad"
;;                   :lastname "panovic"
;;                   :active true
;;                   :email "nenadpann@gmail.com"
;;                   :status "rated"
;;                   :job-id 1},
;;                  {:id 3
;;                   :firstname "arsenije"
;;                   :lastname "pavlovic"
;;                   :active true
;;                   :email "arseenijee00@gmail.com"
;;                   :status "unrated"
;;                   :job-id 1}])

;; (def jobs [{:id 1
;;             :name "C# Junior Developer"
;;             :active true
;;             :positions 1}])

;; (def qualifications [{:id 1
;;                       :name "C# Test"},
;;                      {:id 2
;;                       :name "Education"},
;;                      {:id 3
;;                       :name "Abstract thinking"}])

;; (def criteria [{:job-id 1
;;                 :qualification-id 1
;;                 :ponder 0.5},
;;                {:job-id 1
;;                 :qualification-id 2
;;                 :ponder 0.3},
;;                {:job-id 1
;;                 :qualification-id 3
;;                 :ponder 0.2}])

;; (def ratings [{:id 1
;;                :candidate-id 1
;;                :job-id 1
;;                :qualification-id 1
;;                :value 10},
;;               {:id 2
;;                :candidate-id 1
;;                :job-id 1
;;                :qualification-id 2
;;                :value 10},
;;               {:id 3
;;                :candidate-id 1
;;                :job-id 1
;;                :qualification-id 3
;;                :value 10},

;;               {:id 4
;;                :candidate-id 2
;;                :job-id 1
;;                :qualification-id 1
;;                :value 8},
;;               {:id 5
;;                :candidate-id 2
;;                :job-id 1
;;                :qualification-id 2
;;                :value 9.1},
;;               {:id 6
;;                :candidate-id 2
;;                :job-id 1
;;                :qualification-id 3
;;                :value 7}])


;***********************************************************
;          UCITAVANJE PERZISTENTNIH PODATAKA IZ BAZE
;***********************************************************

(def candidates (repository/get-candidates))
(def jobs (repository/get-jobs))
(def qualifications (repository/get-qualifications))
(def criteria (repository/get-criteria))
(def ratings (repository/get-ratings))

;***********************************************************
;                     OSNOVNE FUNKCIJE 
;***********************************************************

(defn get-candidate [candidate-id]
  (into {} (filter (fn [candidate] (= (:id candidate) candidate-id)) candidates)))

(defn get-ratings
  [candidate-id]
  (into [] (filter (fn [rating] (= (:candidate-id rating) candidate-id)) ratings)))

(get-ratings 1)
(get-ratings 2)


(defn get-candidates
  [job-id]
  (into [] (filter (fn [candidate] (= (:job-id candidate) job-id)) candidates)))

(get-candidates 1)

(defn get-jobs-criteria [job-id]
  (into [] (filter (fn [criteria] (= (:job-id criteria) job-id)) criteria)))

(get-jobs-criteria 1)

(defn get-jobs-name [job-id]
  (:name (into {} (filter (fn [job] (= (:id job) job-id)) jobs))))

(get-jobs-name 1)

(defn get-ponder
  "Returns a ponder value (criteria weight) for a specific criteria"
  [job-id qualification-id]
  (double (:ponder (into {} (filter (fn [criteria] (and (= (:job-id criteria) job-id) (= (:qualification-id criteria) qualification-id))) criteria)))))

(get-ponder 1 1)

;criteria-id = job-id + qualification-id
(defn get-ratings-of-criteria
  "Gets ratings of all candidates for a specific criteria"
  [job-id qualification-id]
  (into [] (filter (fn [rating] (and (= (:job-id rating) job-id) (= (:qualification-id rating) qualification-id))) ratings)))

(get-ratings-of-criteria 1 1)


(defn get-ratings-of-job
  "Gets ratings of all candidates according to all criteria for a specific job"
  [job-id]
  (filter (fn [rating] (= (:job-id rating) job-id)) ratings))

(get-ratings-of-job 1)


;***********************************************************
;                     NORMALIZACIJA 
;***********************************************************

(defn sum-of-all-ratings
  "Calculates sum of ratings of all candidates who have applied for a specific job according to a specific criteria"
  [job-id qualification-id]
  (reduce + (map :value (get-ratings-of-criteria job-id qualification-id))))

(sum-of-all-ratings 1 1)
(sum-of-all-ratings 1 2)
(sum-of-all-ratings 1 3)

(defn get-normalized-ratings-of-criteria
  "Calculates normalized values of ratings according to a specific criteria"
  [job-id qualification-id]
  (into [] (map (fn [row] (assoc row :normalized-value (/ (double (:value row)) (sum-of-all-ratings job-id qualification-id)))) (get-ratings-of-criteria job-id qualification-id))))

(get-normalized-ratings-of-criteria 1 1)
(get-normalized-ratings-of-criteria 1 2)
(get-normalized-ratings-of-criteria 1 3)

(defn get-normalized-rating
  "Returns a rating with added field :normalized-value"
  [rating]
  (into {} (assoc rating :normalized-value (/ (double (:value rating)) (sum-of-all-ratings (:job-id rating) (:qualification-id rating))))))


(defn normalize-job-ratings [job-id]
  "Normalizes all ratings of all candidates for a specific job"
  (let [ratings (get-ratings-of-job job-id)]
    (map (fn [row] (get-normalized-rating row)) ratings)))

(normalize-job-ratings 1)


;***********************************************************
;                     AGREGACIJA OCENA 
;***********************************************************

(defn add-ponder-to-normalized-ratings
  [job-id]
  (let [normalized-ratings (normalize-job-ratings job-id)]
    (into [] (map (fn [row] (assoc row :ponder (get-ponder job-id (:qualification-id row)))) normalized-ratings))))

(add-ponder-to-normalized-ratings 1)

;; (map (fn [row] (* (:normalized-value row) (:ponder row))) (filter (fn [rating] (= (:candidate-id rating) 1)) (add-ponder-to-normalized-ratings 1)))


(defn aggregate-candidate [candidate-id]
  (let [candidate (get-candidate candidate-id)
        ratings (add-ponder-to-normalized-ratings (:job-id candidate))]
    (let [candidates-ratings (filter (fn [rating] (= (:candidate-id rating) candidate-id)) ratings)]
      (assoc candidate :final-score (double (reduce + (map (fn [row] (* (:normalized-value row) (:ponder row))) (filter (fn [rating] (= (:candidate-id rating) candidate-id)) ratings))))))))

(aggregate-candidate 1)
(aggregate-candidate 2)


;***********************************************************
;                   PODRSKA U ODLUCIVANJU 
;***********************************************************


(defn decision-support
  "Applies the method of multi-criteria decision-making and provides
   advices for the most suitable candidates for a specific job"
  [job-id]
  (let [candidates (get-candidates job-id)]
    (into [] (sort-by :final-score (comparator >) (map (fn [row] (aggregate-candidate (:id row))) candidates)))))

(decision-support 1)

(defn round-to-decimal-places [num places]
  (Double/parseDouble (format (str "%." places "f") num)))

(defn selection-advice 
  "Gives advice for the top two most suitable candidates for a specific job"
  [job-id]
  (let [rated-candidates (decision-support job-id)]
    (let [first-candidate (get rated-candidates 0)
          second-candidate (get rated-candidates 1)]
      [(str "Top rated Candidate for " (get-jobs-name job-id) ": " (:firstname first-candidate) " " (:lastname first-candidate) " with final score of " (round-to-decimal-places (:final-score first-candidate) 4))
      (str "You should also consider " (:firstname second-candidate) " " (:lastname second-candidate) " with final score of " (round-to-decimal-places (:final-score second-candidate) 4) " for " (get-jobs-name job-id))])))

(selection-advice 1)


;***********************************************************
;                 NEKI DELOVI STAROG KODA 
;***********************************************************

;; (defn summarize-ratings [candidate-id]
;;  (reduce + (map (fn [row] (:value row)) (get-ratings candidate-id)))) ;daje sumu svih ocena za jednog kandidata

;; (defn summarize-raings1 [candidate-id]
;;   (map (fn [row] (assoc row :normalized-value (/ (double (:value row)) (summarize-ratings candidate-id)))) (get-ratings candidate-id)))
;OVO JE NORMALIZACIJA! --- nije dobro!!!!!
;NE RADI SE SABIRANJE PO SVIM KRITERIJUMIMA ZA JEDNOG KANDIDATA
;NEGO SABIRANJE JEDNOG KRITERIJUMA ZA SVE KANDIDATE!!! 


;; (summarize-raings1 2)

;; (defn summarize-ratings2 [candidate-id] 
;;   (let [candidate (get-candidate candidate-id)]
;;     (assoc candidate :final-score (summarize-ratings candidate-id))))

;; (defn summarize-ratings3 []
;;   (map (fn [row] (summarize-ratings2 (:id row))) candidates)) ;radi za sve kandidate!

;; (summarize-ratings3)
;; (summarize-ratings2 1)

;; (summarize-ratings 1)
;; (summarize-ratings 2)
;; (summarize-ratings 3)



;radi otezanu sumu za jednog kandidata
;; (defn weighted-sum [candidate-id]
;;   (let [candidate (get-candidate candidate-id)]
;;     (let [criteria (get-jobs-criteria (:job-id candidate))
;;           normalized-candidate (summarize-raings1 candidate-id)]
;;       (map (fn [row] (assoc row :weighted-sum (* (:normalized-value row) (get-ponder (:job-id row) (:qualification-id row)))))
;;            normalized-candidate )
;;       )))


;; (weighted-sum 1)

;; (defn aggregated-score [candidate-id]
;;   (let [weighted-candidate (weighted-sum candidate-id)]
;;     (reduce + (map :weighted-sum weighted-candidate))))

;; (aggregated-score 1)
;; (aggregated-score 2)


;; (def niz1 [1 2 3])
;; (def niz2 [3 4 5])
;; ;rezultat - [3 8 15]

;; (into [] (map * niz1 niz2))

(def maps [{:id 1 :number 1}
           {:id 2 :number 100}
           {:id 3 :number 200}])

(map :number maps)


src/clojureselect/repository.clj
(ns clojureselect.repository 
  (:require [next.jdbc :as jdbc-next]
            [next.jdbc.result-set :as rs]))

(def db-spec
  {:dbtype "mysql"
   :dbname "clojureselect"
   :user "root"
   :password "dusan"
   :host "localhost"
   :port 3306})

(def db (jdbc-next/get-datasource db-spec))

(defn get-candidate
  [candidate-id]
  (jdbc-next/execute-one! db
                          ["SELECT * FROM candidates WHERE id = ?" candidate-id] {:builder-fn rs/as-unqualified-maps}))

(defn get-candidates []
  (jdbc-next/execute! db
                                ["SELECT * FROM candidates"] {:builder-fn rs/as-unqualified-maps}))

(defn get-jobs []
  (jdbc-next/execute! db
                          ["SELECT * FROM jobs"] {:builder-fn rs/as-unqualified-maps}))

(defn get-ratings []
  (jdbc-next/execute! db
                          ["SELECT * FROM ratings"] {:builder-fn rs/as-unqualified-maps}))

(defn get-criteria []
  (jdbc-next/execute! db
                          ["SELECT * FROM criteria"] {:builder-fn rs/as-unqualified-maps}))

(defn get-qualifications []
  (jdbc-next/execute! db
                          ["SELECT * FROM qualifications"] {:builder-fn rs/as-unqualified-maps}))

(defn get-employments []
  (jdbc-next/execute! db
                      ["SELECT * FROM employments"] {:builder-fn rs/as-unqualified-maps}))


src/clojureselect/business_logic.clj
(ns clojureselect.business-logic
  (:require [clojureselect.repository :as repository])
  (:gen-class))

;***********************************************************
;            ALATI I METODE VESTACKE INTELIGENCIJE         *
;                I SOFTVERSKOG INZENJERSTVA                *
;                          2023.                           *
;             --------------------------------             *
;                      Clojure Select                      *
;              ~Decision Support Clojure App~              *
;             --------------------------------             *
;                        Dusan Tavic                       *
;                         2023/3801                        *
;***********************************************************



;***********************************************************
;          UCITAVANJE PERZISTENTNIH PODATAKA IZ BAZE
;***********************************************************

(def candidates (repository/get-candidates))
(def jobs (repository/get-jobs))
(def qualifications (repository/get-qualifications))
(def criteria (repository/get-criteria))
(def ratings (repository/get-ratings))

;***********************************************************
;                     OSNOVNE FUNKCIJE 
;***********************************************************

(defn get-candidate 
  "Returns the candidate's object that contains all relevant information, 
   based on the entered id."
  [candidate-id]
  (into {} (filter (fn [candidate] (= (:id candidate) candidate-id)) candidates)))

(defn get-ratings
  "Returns all of the candidate's ratings. 
   Candidates are rated based on selection criteria defined for the job."
  [candidate-id]
  (into [] (filter (fn [rating] (= (:candidate-id rating) candidate-id)) ratings)))

(get-ratings 1)
(get-ratings 2)


(defn get-candidates
  "Returns all candidates who applied for a specific job"
  [job-id]
  (into [] (filter (fn [candidate] (= (:job-id candidate) job-id)) candidates)))

(get-candidates 1)

(defn get-jobs-criteria 
  "Returns all selection criteria for a spec
   ific job"
  [job-id]
  (into [] (filter (fn [criteria] (= (:job-id criteria) job-id)) criteria)))

(get-jobs-criteria 1)

(defn get-jobs-name 
  "Returns the job name based on the entered id"
  [job-id]
  (:name (into {} (filter (fn [job] (= (:id job) job-id)) jobs))))

(get-jobs-name 1)

(defn get-ponder
  "Returns a ponder value (criteria weight) for a specific criteria"
  [job-id qualification-id]
  (if (or (<= job-id 0) (<= qualification-id 0))
    (str "error." " please insert valid id values")
    (double (:ponder (into {} (filter (fn [criteria] (and (= (:job-id criteria) job-id) (= (:qualification-id criteria) qualification-id))) criteria))))))

(get-ponder 1 1)

;criteria-id = job-id + qualification-id
(defn get-ratings-of-criteria
  "Gets ratings of all candidates for a specific criteria"
  [job-id qualification-id]
  (into [] (filter (fn [rating] (and (= (:job-id rating) job-id) (= (:qualification-id rating) qualification-id))) ratings)))

(get-ratings-of-criteria 1 1)


(defn get-ratings-of-job
  "Gets ratings of all candidates according to all criteria for a specific job"
  [job-id]
  (into [] (filter (fn [rating] (= (:job-id rating) job-id)) ratings)))

(get-ratings-of-job 1)


;***********************************************************
;                     NORMALIZACIJA 
;***********************************************************

(defn sum-of-all-ratings
  "Calculates sum of ratings of all candidates who have applied for a specific job according to a specific criteria"
  [job-id qualification-id]
  (reduce + (map :value (get-ratings-of-criteria job-id qualification-id))))

(sum-of-all-ratings 1 1)
(sum-of-all-ratings 1 2)
(sum-of-all-ratings 1 3)

(defn get-normalized-ratings-of-criteria
  "Calculates normalized values of ratings according to a specific criteria"
  [job-id qualification-id]
  (into [] (map (fn [row] (assoc row :normalized-value (/ (double (:value row)) (sum-of-all-ratings job-id qualification-id)))) (get-ratings-of-criteria job-id qualification-id))))

(get-normalized-ratings-of-criteria 1 1)
(get-normalized-ratings-of-criteria 1 2)
(get-normalized-ratings-of-criteria 1 3)

(defn get-normalized-rating
  "Returns a rating with added field :normalized-value"
  [rating]
  (into {} (assoc rating :normalized-value (/ (double (:value rating)) (sum-of-all-ratings (:job-id rating) (:qualification-id rating))))))


(defn normalize-job-ratings 
  "Normalizes all ratings of all candidates for a specific job"
  [job-id] 
  (let [ratings (get-ratings-of-job job-id)]
    (map (fn [row] (get-normalized-rating row)) ratings)))

(normalize-job-ratings 1)


;***********************************************************
;                     AGREGACIJA OCENA 
;***********************************************************

(defn add-ponder-to-normalized-ratings
  "Adds a ponder to the normalized ratings of all candidates who applied for a specific job"
  [job-id]
  (let [normalized-ratings (normalize-job-ratings job-id)]
    (into [] (map (fn [row] (assoc row :ponder (get-ponder job-id (:qualification-id row)))) normalized-ratings))))

(add-ponder-to-normalized-ratings 1)

;; (map (fn [row] (* (:normalized-value row) (:ponder row))) (filter (fn [rating] (= (:candidate-id rating) 1)) (add-ponder-to-normalized-ratings 1)))


(defn aggregate-candidate 
  "Aggregates the overall rating of the candidate by using a weighted sum"
  [candidate-id]
  (let [candidate (get-candidate candidate-id)
        ratings (add-ponder-to-normalized-ratings (:job-id candidate))]
    (let [candidates-ratings (filter (fn [rating] (= (:candidate-id rating) candidate-id)) ratings)]
      (assoc candidate :final-score (double (reduce + (map (fn [row] (* (:normalized-value row) (:ponder row))) (filter (fn [rating] (= (:candidate-id rating) candidate-id)) ratings))))))))

(aggregate-candidate 1)
(aggregate-candidate 2)


;***********************************************************
;                   PODRSKA U ODLUCIVANJU 
;***********************************************************


(defn decision-support
  "Applies the method of multi-criteria decision-making and provides
   advices for the most suitable candidates for a specific job"
  [job-id]
  (let [candidates (get-candidates job-id)]
    (into [] (sort-by :final-score (comparator >) (map (fn [row] (aggregate-candidate (:id row))) candidates)))))

(decision-support 1)

(defn round-to-decimal-places [num places]
  (Double/parseDouble (format (str "%." places "f") num)))

(defn selection-advice 
  "Gives advice for the top two most suitable candidates for a specific job"
  [job-id]
  (let [rated-candidates (decision-support job-id)]
    (let [first-candidate (get rated-candidates 0)
          second-candidate (get rated-candidates 1)]
      [(str "Top rated Candidate for " (get-jobs-name job-id) ": " (:firstname first-candidate) " " (:lastname first-candidate) " with final score of " (round-to-decimal-places (:final-score first-candidate) 4))
      (str "You should also consider " (:firstname second-candidate) " " (:lastname second-candidate) " with final score of " (round-to-decimal-places (:final-score second-candidate) 4) " for " (get-jobs-name job-id))])))

(selection-advice 1)


;***********************************************************
;        AHP - ODREDJIVANJE VAZNOSTI KRITERIJUMA
;***********************************************************
; u okviru ahp-ponders nalaze se preference donosilaca odluka 


(def ahp-ponders [{:job-id 1
                   :qualification-id-base 1
                   :qualification-id-reference 2 
                    :position [0,1]
                   :significance 3},
                  {:job-id 1
                   :qualification-id-base 1
                   :qualification-id-reference 3
                   :position [0,2]
                   :significance 2},
                  {:job-id 1
                   :qualification-id-base 2
                   :qualification-id-reference 3
                   :position [1,2]
                   :significance 0.5}])

(defn get-ahp-ponders
  "Returns ahp ponders for a specific job"
  [job-id]
  (into [] (filter (fn [row] (= (:job-id row) job-id)) ahp-ponders)))

(defn create-matrix
  "Creates a matrix with specific number of rows and columns, initially filled with ones"
  ([rows columns]
  (vec (for [x (range rows)]
         (vec (repeat columns 1)))))
  ([rows-and-cols]
   (vec (for [x (range rows-and-cols)]
          (vec (repeat rows-and-cols 1))))
   ))

;za sada samo za jedan kriterijum radi 
(defn create-ahp-matrix
  "Creates AHP matrix of assessments for a specific job"
  [job-id]
  (let [ahp-ponders (get-ahp-ponders job-id)
        rows-cols-count (count ahp-ponders)
        init-matrix (create-matrix rows-cols-count)] 
   (assoc-in init-matrix [(get (:position (get ahp-ponders 0)) 0) (get (:position (get ahp-ponders 0)) 1)] (:significance (get ahp-ponders 0)))
  )) 

(create-ahp-matrix 1)



;***********************************************************
;                 SIMULACIJA PODATAKA U BAZI 
;***********************************************************
; koristi se iskljucivo za potrebe testiranja funkcija

(def candidates-sim [{:id 1
                    :firstname "mario"
                    :lastname "tavic"
                    :active true
                    :email "dusantavic1@gmail.com"
                    :status "rated"
                    :job-id 1},
                   {:id 2
                    :firstname "nenad"
                    :lastname "panovic"
                    :active true
                    :email "nenadpann@gmail.com"
                    :status "rated"
                    :job-id 1},
                   {:id 3
                    :firstname "arsenije"
                    :lastname "pavlovic"
                    :active true
                    :email "arseenijee00@gmail.com"
                    :status "unrated"
                    :job-id 1}])

(def jobs-sim [{:id 1
              :name "C# Junior Developer"
              :active true
              :positions 1}])

(def qualifications-sim [{:id 1
                        :name "C# Test"},
                       {:id 2
                        :name "Education"},
                       {:id 3
                        :name "Abstract thinking"}])

(def criteria-sim [{:job-id 1
                  :qualification-id 1
                  :ponder 0.5},
                 {:job-id 1
                  :qualification-id 2
                  :ponder 0.3},
                 {:job-id 1
                  :qualification-id 3
                  :ponder 0.2}])


(def ratings-sim [{:id 1
                 :candidate-id 1
                 :job-id 1
                 :qualification-id 1
                 :value 10},
                {:id 2
                 :candidate-id 1
                 :job-id 1
                 :qualification-id 2
                 :value 10},
                {:id 3
                 :candidate-id 1
                 :job-id 1
                 :qualification-id 3
                 :value 10},

                {:id 4
                 :candidate-id 2
                 :job-id 1
                 :qualification-id 1
                 :value 8},
                {:id 5
                 :candidate-id 2
                 :job-id 1
                 :qualification-id 2
                 :value 9.1},
                {:id 6
                 :candidate-id 2
                 :job-id 1
                 :qualification-id 3
                 :value 7}])



